File: wand.pl
#!/usr/bin/perl -w
use strict;
use warnings;
$| = 1;
use IPC::Open3;
use File::Spec;
use Symbol qw(gensym);

my $counter = 0;
my $current_nibble = '';
my $total_counter = 0;
my $current_reading = '';
my $good_readings = 0;
my $last_space_time = 0;
my $last_duty_cycle = 0;

my $pid;
my $file = shift; # used to test captures of ir-ctl output
if (defined $file and $file) {
  open INPUT, "<$file" or die $!;
} else {
  my $irctlcmd = "ir-ctl -r";
  open(NULL, ">", File::Spec->devnull);
  $pid = open3(gensym, \*INPUT, ">&NULL", "$irctlcmd");
}

my $DURATION_CUTOFF =  shift || 0.3481; # best results through testing

while( <INPUT> ) { 
  if (m/^(pulse|space|timeout) (\d+)$/) {

    my ($what, $time) = ($1, $2);

    if ($what eq 'pulse') {
      $counter++;
      $total_counter++;

      if ($last_space_time > 0) {
        my $duty_cycle = $last_duty_cycle = $last_space_time + $time;
        my $duration = $time/$duty_cycle;
        #warn int($duration*100) . " " ;
        if ($duration < $DURATION_CUTOFF) {
          $current_nibble .= 0;
        } else {
          $current_nibble .= 1;
        }
      } else {
        # This must be the first pulse in the reading, so assume it's a 0.
        # ir-ctl starts with a pulse, not a space.
        $current_nibble .= 0;
      }
    } elsif ($what eq 'space') {
      $last_space_time = $time;
    } elsif ($what eq 'timeout') {
      if (($total_counter > 0) and ($total_counter < 56)) {
        warn "Reading was short: $total_counter bits\n";
      }
      $counter = $total_counter = 0;
      $current_nibble = '';
      $current_reading = '';
    }
  } else {
    warn "Didn't match anything: $_";
  }

  if ($counter == 4) {
    my $nibble = sprintf("%x", oct("0b$current_nibble"));
    $current_reading .= $nibble;
    $current_nibble = '';
    $counter = 0;
  }
  if ($total_counter == 56) {
    handle_reading($current_reading);
    $current_reading = '';
    $total_counter = 0;
  }

}
warn "Closing\n";
warn "Good readings: $good_readings\n";
if (defined $pid and $pid) {
  close(NULL);
  waitpid($pid, 0);
} else {
  close(INPUT);
}

sub handle_reading {
  my ($input) = @_;
  #warn "Got a whole reading: $input\n";
  if ($input =~ m/(..)(......)(......)/) {
    my $zeroes = $1;
    my $id = $2;
    my $motion = $3; 
    if ($zeroes ne '00') {
      warn "Reading $input didn't start with 00";
      return;
    } 
    $good_readings++;
    print "ID:$id\n";
  } else {
    warn "Didn't get a good ID from $input";
  }
}

File: wand.js
const fs = require('fs');
const readline = require('readline');
const util = require('util');
var spawn = require('child_process').spawn;
var irdata = spawn('/usr/bin/ir-ctl', ['-r']);
//var irdata = spawn('/bin/cat', ['./new.txt']);

console.log("Started");

var ir_counter = 0;
var ir_current_nibble = '';
var ir_total_counter = 0;
var ir_current_reading = '';
var ir_last_space_time = 0;
var ir_last_duty_cycle = 0;
const ir_duration_cutoff = 0.3481;

var lines_buffer = [];

irdata.stdout.on('data', function(data) {
  var str = data.toString();
  var lines = str.split(/\n/g);
  lines_buffer.push(lines);
});

function waitForLines() {
  if (lines_buffer.length > 0) {
    var lines = lines_buffer.shift();
    for (var i=0; i < lines.length; i++) {
      handleLine(lines[i]);
    }
  } else {
    //console.log('waiting');
  }
}

setInterval(waitForLines, 100);

function handleLine(line) {
  var m;
  if (m = line.match(/(pulse|space|timeout) (\d+)/)) {
    var what = m[1];
    var timing = parseInt(m[2]);
    //console.log('What: ' + what + ' -> ' + timing);

    switch(what) {
      case 'pulse': 
        ir_counter++;
        ir_total_counter++;
        if (ir_last_space_time > 0) {
          var duty_cycle = ir_last_duty_cycle = ir_last_space_time + timing;
          var duration = timing/duty_cycle;
          if (duration < ir_duration_cutoff) {
            ir_current_nibble += 0;
          } else {
            ir_current_nibble += 1;
          }
        } else {
          // First pulse in the reading, so assume 0
          ir_current_nibble += 0;
        }
        //console.log(ir_counter + ' ' + ir_total_counter + ' ' + ir_current_nibble + ' ' + timing + ' ' + duty_cycle);
        break;
      case 'space': 
        ir_last_space_time = timing
        break;
      case 'timeout': 
        if ((ir_total_counter > 0) && (ir_total_counter < 56)) {
          console.log("Reading was short: " + ir_total_counter + " bits");
        } else if (ir_total_counter > 56) {
          console.log("Reading was too long: " + ir_total_counter + " bits");
        }
        ir_counter = ir_total_counter = 0;
        ir_current_nibble = ir_current_reading = '';
        break;
    }
    if (ir_counter == 4) {
      nibble = parseInt(ir_current_nibble, 2).toString(16);
      ir_current_reading += nibble;
      ir_current_nibble = '';
      ir_counter = 0;
    }
    if (ir_total_counter == 56) {
      var result = ir_current_reading;
      ir_current_reading = '';
      ir_total_counter = 0;
      handleResult(result);
    }
  }
}

function handleResult(reading) {
  console.log('Raw reading: ' + reading);
  if (m = reading.match(/(..)(......)(......)/)) {
    var zeroes = m[1];
    var id = m[2];
    var motion = m[3];
    if (zeroes != '00') {
      console.log('Reading did not start with 00');
      return;
    }
    console.log('ID: ' + id);
    return id;
  }
}

irdata.stderr.on('data', function(data) {
  console.log('stderr: ' + data.toString());
});

irdata.on('exit', function (code) {
  console.log('child process exited with code ' + code);
});
